### PR TITLE
support kubernetes version with expiration date

### LINKED ISSUE #1325: Please support Kubernetes version expiration date
**What would you like to be added**:
When managing could profiles and Kubernetes versions, we tend to drop versions with security vulnerabilities rather instantly, replacing them with the latest safe one. This causes issues with stakeholders that first want to validate the update process, but have no means anymore to create an "old" cluster and run through the update process. While this is more of a process issue (not strictly relevant to Gardener code), it should be possible to specify expiration dates on Kubernetes versions and the API server should reject shoot resources with expired Kubernetes versions (this, on the other hand, would be relevant to Gardener code).

**Why is this needed**:
To have enough time to validate Kubernetes version updates before they are forcefully executed.

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**What this PR does / why we need it**:
Adding the ability to specify expiration dates on Kubernetes Versions in the CloudProfile.

- Shoots with expired Kubernetes Versions  will be updated to the latest available patch version during the maintenance time window
- Shoots that run a Kubernetes Version that does not exist any more in the CloudProfile will be updated to the latest available patch version during the maintenance time window
- Expiration Date on latest K8 Version is not allowed

**Also**
- Adding the Kubernetes Version Maintenance to the Maintenance Integration Tests
- Removing the .test-dev/cmd integration test to avoid duplicated effort -> test machinery also using the test/integration test 

**Which issue(s) this PR fixes**:
Fixes #1325

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
If you opted-out of the automatic Kubernetes patch version update the shoot Kubernetes version won't be updated unless the expiration date specified by the administrator for the used version expires. The expiration date for the respective Kubernetes versions can be found in the referenced `CloudProfile` in the `.spec.<cloud>.constraints.kubernetes[].offeredVersions[].expirationDate` fields. Example: The `CloudProfile` specifies `(1.14.2, 2019-08-08T14:00:00Z)` and `(1.14.5)` as possible versions for `1.14`. A shoot that is opted-out of automatic updates uses `(1.14.2)`. In its maintenance time window it won't be updated to `1.14.5` before `2019-08-08T14:00:00Z`.
```
```noteworthy operator
The `.spec.<cloud>.constraints.kubernetes[].versions` field in the `CloudProfile` is deprecated and will be removed in the future. It is now replaced with a list of offered versions (`.spec.<cloud>.constraints.kubernetes[].offeredVersions`). Each version can have an `expirationDate` which specifies the time after all shoots that are using this version and opted-out of automatic Kubernetes patch version updates will get forcefully updated to the latest Kubernetes patch version for the used `<major>.<minor>` version. Example: The `CloudProfile` specifies `(1.14.2, 2019-08-08T14:00:00Z)` and `(1.14.5)` as possible versions for `1.14`. A shoot that is opted-out of automatic updates uses `(1.14.2)`. In its maintenance time window it won't be updated to `1.14.5` before `2019-08-08T14:00:00Z`.
```

### CODE REVIEW

File: pkg/apis/garden/types.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ type AWSConstraints struct {
 type MachineImage struct {
 	// Name is the name of the image.
 	Name string
-	// Versions contains versions and expiration dates of the machine image
+	// OfferedVersions contains versions and expiration dates of the machine image
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]: This should not be changed.

File: pkg/apis/garden/types.go [hunk 2]
```diff
@@ -279,8 +279,18 @@ type DNSProviderConstraint struct {
...
 // KubernetesConstraints contains constraints regarding allowed values of the 'kubernetes' block in the Shoot specification.
 type KubernetesConstraints struct {
-	// Versions is the list of allowed Kubernetes versions for Shoot clusters (e.g., 1.13.1).
-	Versions []string
+	// OfferedVersions is the list of allowed Kubernetes versions with optional expiration dates for Shoot clusters
+	OfferedVersions []OfferedVersion
+}
+
+// OfferedVersion contains the version code and optional expiration date for a kubernetes version
+type OfferedVersion struct {
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]: Can you call it `KubernetesVersion`?

File: pkg/apis/garden/v1beta1/types.go [hunk 3]
```diff
@@ -114,7 +114,7 @@ type MachineImage struct {
 	// DEPRECATED: This field will be removed in a future version.
 	// +optional
 	Version string `json:"version,omitempty"`
-	// Versions contains versions and expiration dates of the machine image
+	// OfferedVersions contains versions and expiration dates of the machine image
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]: This should not be changed.

File: pkg/apis/garden/v1beta1/types.go [hunk 4]
```diff
@@ -299,8 +299,22 @@ type DNSProviderConstraint struct {
 
 // KubernetesConstraints contains constraints regarding allowed values of the 'kubernetes' block in the Shoot specification.
 type KubernetesConstraints struct {
-	// Versions is the list of allowed Kubernetes versions for Shoot clusters (e.g., 1.13.1).
+	// DEPRECATED: This field will be removed in a future version.
+	// OfferedVersions is the list of allowed Kubernetes versions for Shoot clusters (e.g., 1.13.1).
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
	// OfferedVersions is the list of allowed Kubernetes versions for Shoot clusters (e.g., 1.13.1).
	// +optional
```

File: pkg/apis/garden/v1beta1/types.go [hunk 5]
```diff
@@ -299,8 +299,22 @@ type DNSProviderConstraint struct {
...
 	Versions []string `json:"versions"`
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
	Versions []string `json:"versions,omitempty"`
```

File: pkg/apis/garden/v1beta1/types.go [hunk 6]
```diff
@@ -299,8 +299,22 @@ type DNSProviderConstraint struct {
...
+	// OfferedVersions is the list of allowed Kubernetes versions with optional expiration dates for Shoot clusters
+	OfferedVersions []OfferedVersion `json:"offeredVersions"`
+}
+
+// OfferedVersion contains the version code and optional expiration date for a kubernetes version
+type OfferedVersion struct {
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
type KubernetesVersion struct {
```

File: pkg/apis/garden/v1beta1/helper/helpers.go [hunk 7]
```diff
@@ -487,73 +487,129 @@ func DetermineLatestKubernetesPatchVersion(cloudProfile gardenv1beta1.CloudProfi
...
-	for _, version := range versions {
-		ok, err := utils.CompareVersions(version, operator, currentVersion)
-		if err != nil {
-			return false, []string{}, err
-		}
-		if version != currentVersion && ok {
-			newerVersions = append(newerVersions, version)
-		}
+// SetKubernetesVersions sets the Kubernetes Versions to the CloudProfile
+func SetKubernetesVersions(profile *gardenv1beta1.CloudProfile, offeredVersions []gardenv1beta1.OfferedVersion, versions []string) error {
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/helper/helpers.go]: I don't see this function used anywhere, can we remove it?

File: pkg/apis/garden/validation/validation.go [hunk 8]
```diff
@@ -273,14 +273,23 @@ func ValidateCloudProfileSpec(spec *garden.CloudProfileSpec, fldPath *field.Path
...
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, latestKubernetesVersion, "failed to determine latest kubernetes version from cloud profile"))
+	}
+
+	if latestKubernetesVersion.ExpirationDate != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("offeredVersions[]").Child("expirationDate"), latestKubernetesVersion.ExpirationDate, fmt.Sprintf("expiration date of latest kubernetes version ('%s') must not be set", latestKubernetesVersion.Version)))
 	}
 
 	r, _ := regexp.Compile(`^([0-9]+\.){2}[0-9]+$`)
-	for i, version := range kubernetes.Versions {
```
Comment (rfranzke) [inline pkg/apis/garden/validation/validation.go:280]: I think we should still validate that these versions are correct.

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 9]
```diff
@@ -220,6 +219,37 @@ func (c *defaultMaintenanceControl) Maintain(shootObj *gardenv1beta1.Shoot, key
...
+	return nil, nil
+}
+
+func shouldKubernetesVersionBeUpdated(shoot *gardenv1beta1.Shoot, profile *gardenv1beta1.CloudProfile) (bool, error) {
+	versionExistsInCloudProfile, offeredVersion, err := helper.KubernetesVersionExistsInCloudProfile(*profile, shoot.Spec.Kubernetes.Version)
+	if err != nil {
+		return false, err
+	}
+
+	if !versionExistsInCloudProfile || shoot.Spec.Maintenance.AutoUpdate.KubernetesVersion || ExpirationDateExpired(offeredVersion.ExpirationDate) {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]:
```suggestion
	return !versionExistsInCloudProfile || shoot.Spec.Maintenance.AutoUpdate.KubernetesVersion || ExpirationDateExpired(offeredVersion.ExpirationDate, nil
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/garden/validation/validation.go (hunk 8 above)
Comment (danielfoehrKn) [reply pkg/apis/garden/validation/validation.go:280]: This is the validation for the internal api. In the internal version, the versions field does not exists. the offeredVersions contains all "versions" and "offeredVersions" from the external representation.

File: pkg/apis/garden/v1beta1/helper/helpers.go (hunk 7 above)
Comment (danielfoehrKn) [reply pkg/apis/garden/v1beta1/helper/helpers.go]: its being used couple times in the integration tests :)

File: pkg/apis/garden/validation/validation.go (hunk 8 above)
Comment (rfranzke) [reply pkg/apis/garden/validation/validation.go:280]: Sure, thank you, thanks.

**Comment by rfranzke** [comment]
The tests are failing, can you check, please?

**Review by rfranzke — DISMISSED** [review]

**Pull request merged**